ldms/src/sampler/lustre_mdc/lustre_mdc.c

```diff
@@ -0,0 +1,288 @@
+/* -*- c-basic-offset: 8 -*- */
+/* Copyright 2021 Lawrence Livermore National Security, LLC
+ * See the top-level COPYING file for details.
+ * Copyright 2022 NTESS.
+ *
+ * SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+ */
+#include <limits.h>
+#include <string.h>
+#include <dirent.h>
+#include <coll/rbt.h>
+#include <sys/queue.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "config.h"
+#include "lustre_mdc.h"
+#include "lustre_mdc_general.h"
+#include "jobid_helper.h"
+
+#define _GNU_SOURCE
+
+#define MDC_PATH "/proc/fs/lustre/mdc"
+#define TIMING_SEARCH_PATH "/sys/kernel/debug/lustre/mdc/"
+
+static struct comp_id_data cid;
+
+ldmsd_msg_log_f log_fn;
+char producer_name[LDMS_PRODUCER_NAME_MAX];
+
+/* red-black tree root for mdcs */
+static struct rbt mdc_tree;
+
+static struct base_auth auth;
+static int mdc_timing = 1;
+
+struct mdc_data {
+	char *fs_name;
+	char *name;
+	char *path;
+	char *stats_path; /* operations: /proc/fs/lustre/mdc/ * /md_stats */
+	char *timing_path; /* timings: /sys/kernel/debug/lustre/mdc/ * /stats */
+	ldms_set_t general_metric_set; /* a pointer */
+	struct rbn mdc_tree_node;
+};
+
+static int string_comparator(void *a, const void *b)
+{
+	return strcmp((char *)a, (char *)b);
+}
+
+static struct mdc_data *mdc_create(const char *mdc_name, const char *basedir)
+{
+	struct mdc_data *mdc;
+	char path_tmp[PATH_MAX]; /* TODO: move large stack allocation to heap */
+	char *state;
+
+	log_fn(LDMSD_LDEBUG, SAMP" mdc_create() %s from %s\n",
+	       mdc_name, basedir);
+	mdc = calloc(1, sizeof(*mdc));
+	if (mdc == NULL)
+		goto out1;
+	mdc->name = strdup(mdc_name);
+	if (mdc->name == NULL)
+		goto out2;
+	snprintf(path_tmp, PATH_MAX, "%s/%s", basedir, mdc_name);
+	mdc->path = strdup(path_tmp);
+	if (mdc->path == NULL)
+		goto out3;
+	snprintf(path_tmp, PATH_MAX, "%s/md_stats", mdc->path);
+	mdc->stats_path = strdup(path_tmp);
```

Comment:
I think this naming is going to be very confusing in the future. We should not rename "md_stats" to "stats" and rename "stats" to "timing". "stats" is (almost?) always a file with a standardized format.

In other words, I think we would be better off with "mdc->md_stats_path" and "mdc->stats_path".

Reply:
I can split the set, but I did not initially because:
 * there's no way to specify that I want set A at interval one and set B at interval two.
 * it adds transmission overhead
 * both sets carry the same metadata including device for the metrics
 
I can certainly adjust the schema name dependent on its content (mdc_ops or mdc_ops_timing) so there is no backend heartache based on config params.

Until we bump to a major version change and rename all the samplers, i think we only need/want the _sampler suffix on samplers which need it to avoid conflicts by having libxxx_sampler vs libxxx (like dcgm). but that's just my opinion.

Reply:
> stats files seems to be standard on in the sense of a column format rather than a file format, where content varies wildly row to row and there are already examples of abusing the column format.

I'm pretty sure that "stats" is generated by the same function in Lustre regardless of its location. Where do you think you see it changing the column format?

> I will look into the internal renaming of things so that variable names line up with file leaf names in the specific mdc context. At the user level, 'stats' is apparently meaningless in the sense that it's a catch-all file name that means anything the context (lustre subsystem) needs it to mean. To me, 'timing' is much more specific and meaningful.

I'm talking about in the code. Anyone working on the lustre sampler code would presumably need to have some familiarity with lustre naming. And no on familiar with lustre calls will recognized "timing" as a specific in lustre. A "stats" file in lustre is a well-known format in the lustre world.

Reply:
I have renamed vars/members  in c to correspond to basename of the file being read and adjusted schema name depending on content.